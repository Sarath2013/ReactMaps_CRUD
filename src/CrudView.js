import React, { PureComponent } from 'react';
import LocationDetials from './LocationDetials'
import AddModal from './AddModal';
import axios from 'axios';
import PropTypes from 'prop-types';

class CrudView extends PureComponent {
  constructor(props) {
    super(props);
    this.state = { showAddModal: false, modalName: '', curLocation: {}, activeEditIndex: undefined, hasError: false };
    this.updateLocations = this.updateLocations.bind(this);
    this.loadMarkers = this.loadMarkers.bind(this);
    this.toggleAddOpen = this.toggleAddOpen.bind(this);
    this.toggleEditOpen = this.toggleEditOpen.bind(this);
    this.deleteLocation = this.deleteLocation.bind(this);
    this.toggleAddEditComplete = this.toggleAddEditComplete.bind(this);
    this.toggleAddEditClose = this.toggleAddEditClose.bind(this);
    this.viewMaker = this.viewMaker.bind(this);
  }

  loadMarkers() {
    return (this.props.locations.length > 0) ? (this.props.locations.map((location, index) => {
      return <div className="col-md-6 col-sm-12 col-xs-12 loc-details" key={index}>
        <LocationDetials location={location} />
        <div className='loc-actions'>
          <div className="view-marker"><button className="btn btn-default" onClick={() => this.viewMaker(index)}>View Marker</button></div>
          <span><button className="btn btn-default" onClick={() => this.toggleEditOpen(index)}>Edit</button></span>
          <span>or</span>
          <span><button className="btn btn-default" onClick={() => this.deleteLocation(index)}>Delete</button></span>
          {(this.state.hasError) ?
            <div className="error-msg">Something went wrong. Please try agian later.</div> : null}
        </div>
      </div>
    })) : <div className="no-markers"><h6>No markers to show. Please add markers</h6></div>;
  }

  updateLocations() {
    this.props.update(this.props.locations);
  }

  toggleAddOpen() {
    this.setState({ showAddModal: !this.state.showAddModal, modalName: 'Add', curLocation: { name: '' }, activeEditIndex: undefined });
  }

  toggleAddEditComplete(data) {
    let locations, activeValues = data;
    if (this.state.activeEditIndex >= 0) {
      locations = [...this.props.locations];
      locations[this.state.activeEditIndex] = data;
    }
    else
      locations = [...this.props.locations, data];
    this.setState({ showAddModal: !this.state.showAddModal, curLocation: {}, activeEditIndex: undefined });
    this.props.update(locations, activeValues);
  }

  toggleAddEditClose(data) {
    this.setState({ showAddModal: !this.state.showAddModal, curLocation: {}, activeEditIndex: undefined });
  }

  viewMaker(index) {
    let curLocation = this.props.locations[index];
    this.props.update(this.props.locations, curLocation);
  }

  toggleEditOpen(index) {
    let curLocation = this.props.locations[index];
    this.setState({ showAddModal: !this.state.showAddModal, modalName: 'Edit', curLocation, activeEditIndex: index });
  }

  deleteLocation(index) {
    let locations = [...this.props.locations];
    locations.splice(index, 1);
    axios.delete(`http://localhost:4000/deleteMarker/${this.props.locations[index]._id}`, { headers: { "Access-Control-Allow-Origin": "*" } })
      .then(res => {
        this.props.update(locations);
      })
      .catch(error => {
        console.log(error);
        this.setState({ hasError: true });
      })
  }

  render() {
    return (
      <div className="contianer">
        <div className="row">
          <div className="col-md-4">
            <button className="btn btn-primary" onClick={this.toggleAddOpen} >Add Map</button>
          </div>
        </div>
        <hr />
        <div className="row">
          {this.loadMarkers()}
        </div>
        {
          this.state.showAddModal ? (<div className="add-modal-container"><AddModal toggle={this.toggleAddEditComplete} toggleClose={this.toggleAddEditClose} modalName={this.state.modalName} curLocation={this.state.curLocation} locations={this.props.locations} />
            <div className="modal-backdrop show"></div></div>
          ) : null
        }
      </div>
    );
  }
}

CrudView.propTypes = {
  locations: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    lat: PropTypes.number.isRequired,
    lng: PropTypes.number.isRequired
  })).isRequired,
  update: PropTypes.func.isRequired
}

export default CrudView;