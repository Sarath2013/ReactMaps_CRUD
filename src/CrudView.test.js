import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import CrudView from './CrudView';

const props = {
  locations: [{
    name: 'Berlin',
    lat: 52.5200,
    lng: 13.4050
  },
  {
    name: 'Munich',
    lat: 48.1351,
    lng: 11.5820
  }],
  update: jest.fn()
};

let wrapper;

describe("Rendering CrudView", () => {
  beforeAll(() => {
  });

  afterAll(() => {
    wrapper = null;
  });

  test('renders snapshot', () => {
    wrapper = shallow(<CrudView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('check markers count', () => {
    wrapper = mount(<CrudView {...props} />);
    wrapper.update();
    expect(wrapper.props().locations.length).toBe(2);
    expect(wrapper.props().locations).toEqual([{
      name: 'Berlin',
      lat: 52.52,
      lng: 13.405
    },
    {
      name: 'Munich',
      lat: 48.1351,
      lng: 11.582
    }]);
    expect(wrapper.find('button.btn.btn-primary').text()).toBe('Add Map');
    const locations = wrapper.find('.loc-details');
    expect(locations.length).toBe(2);
    const details = locations.get(0);
    expect(details.props.children[0].props.location.name).toBe('Berlin');
    expect(details.props.children[0].props.location.lat).toBe(52.52);
    expect(details.props.children[0].props.location.lng).toBe(13.405);
    expect(locations.get(1).props.children[0].props.location.name).toBe('Munich');
    expect(locations.get(1).props.children[0].props.location.lat).toBe(48.1351);
    expect(locations.get(1).props.children[0].props.location.lng).toBe(11.582);
    const actionBtns = wrapper.find('button.btn.btn-default');
    expect(actionBtns.get(0).props.children).toBe('View Marker');
    expect(actionBtns.get(1).props.children).toBe('Edit');
    expect(actionBtns.get(2).props.children).toBe('Delete');
    wrapper.unmount();
  });
});


