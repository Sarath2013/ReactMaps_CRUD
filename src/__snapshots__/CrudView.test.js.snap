// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering CrudView renders snapshot 1`] = `
<div
  className="contianer"
>
  <div
    className="row"
  >
    <div
      className="col-md-4"
    >
      <button
        className="btn btn-primary"
        onClick={[Function]}
      >
        Add Map
      </button>
    </div>
  </div>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-md-6 col-sm-12 col-xs-12 loc-details"
      key="0"
    >
      <LocationDetials
        location={
          Object {
            "lat": 52.52,
            "lng": 13.405,
            "name": "Berlin",
          }
        }
      />
      <div
        className="loc-actions"
      >
        <div
          className="view-marker"
        >
          <button
            className="btn btn-default"
            onClick={[Function]}
          >
            View Marker
          </button>
        </div>
        <span>
          <button
            className="btn btn-default"
            onClick={[Function]}
          >
            Edit
          </button>
        </span>
        <span>
          or
        </span>
        <span>
          <button
            className="btn btn-default"
            onClick={[Function]}
          >
            Delete
          </button>
        </span>
      </div>
    </div>
    <div
      className="col-md-6 col-sm-12 col-xs-12 loc-details"
      key="1"
    >
      <LocationDetials
        location={
          Object {
            "lat": 48.1351,
            "lng": 11.582,
            "name": "Munich",
          }
        }
      />
      <div
        className="loc-actions"
      >
        <div
          className="view-marker"
        >
          <button
            className="btn btn-default"
            onClick={[Function]}
          >
            View Marker
          </button>
        </div>
        <span>
          <button
            className="btn btn-default"
            onClick={[Function]}
          >
            Edit
          </button>
        </span>
        <span>
          or
        </span>
        <span>
          <button
            className="btn btn-default"
            onClick={[Function]}
          >
            Delete
          </button>
        </span>
      </div>
    </div>
  </div>
</div>
`;
